### DISCUSSION TITLE
Dynamic Routes with Next Export (and say - Vercel)

### DISCUSSION THREAD

**Discussion by colmanhumphrey** [vercel/next.js#28216] [Help] [answered]
I know some similar things have come up, but I haven't yet been able to solve this, I may just be doing something a bit daft.

## Setup: Dynamic Routes

Basically I have dynamic routes, from pages of the following form:
- `/pages/apps/[user]/index.tsx`
- `/pages/apps/[user]/[id].tsx`

Both files then take the incoming `user` and `id` query parameters and fetch some data.

This works great locally with `next dev`, and I presume would work fine with `next start` on deployment. 

## Issue: Doesn't work on export of static files

So I think ideally this could be built with `next build && next export` or similar and work. From discussions and issues raised here, I know it's not really possible for this to work out of the box everywhere without switching to a full SPA model. But I think this should be doable at least in some specific circumstances. I'm very happy to host directly on Vercel too, and it seems that others have had a good experience with exactly this problem. However, even on Vercel, I'm getting 404s for these pages. All non-dynamic pages are working, and the dynamic pages work if linked to from another page.

I won't know all the routes etc ahead of time, so some solutions available with `getStaticPaths` etc probably won't work.

Two notes from @timneutkens:
- [Here, he discusses that `next export` would require proxy-level routing to solve the general dynamic cases](https://github.com/vercel/next.js/discussions/11956#discussioncomment-4965), but that Vercel (previously ZEIT) should have this working out of the box
- I think he [makes the same general point here](https://github.com/vercel/next.js/discussions/12121#discussioncomment-6266), that Vercel should have this working out of the box

[This previous NextJS](https://github.com/vercel/next.js/discussions/11178#discussioncomment-4802) discussion is someone who has this all working on Vercel, but is trying to get it to work on Netlify - and later [an example solution](https://github.com/jericopulvera/netlify-rewrite-test) is given for Netlify, using wildcard redirects.

Hopefully this is actually possible and I'm not misinterpreting the various discussions - and ideally I'm just doing something dumb and/or my route names are slightly wrong!

**Comment by colmanhumphrey** [comment, marked as answer]
After changing the build step to `next build` instead of `next build && next export`, it all works. But... Is this actually serving the site statically? Or now is it running on a server?

**Reply by leerob** [reply]
Are you using Vercel? I can only speak for us, but your static assets / pages are still served statically using `next build`. There's no advantage of using `next export` on Vercel 👍

**Reply by colmanhumphrey** [reply]
Ahh gotcha, thats great! Just on a final note - if that's the case, why does using `next export` break the dynamic features even on Vercel?

**Reply by leerob** [reply]
`next export` doesn't _break_ anything -> it's purposely limited due to a lack of a server. For example, you can't look at the incoming request to check for the `Accept-Lang` header when you don't have a server, so i18n routing isn't supported. Similar story with API routes, on-demand image optimization, etc.

Next.js (and using the standard `next build`) will try to [automatically optimize](https://nextjs.org/docs/advanced-features/automatic-static-optimization) for static, even if you don't explicitly opt-in (like with `getStaticProps`).

**Reply by colmanhumphrey** [reply]
Ok sounds good! So basically when Tim in the links was suggesting Vercel supports the required proxy-level routing out of the box for `next export`, he wasn't so much being literal, but really just expressing that on Vercel, `next bulid` gives pretty much the same benefits as export in terms of static site building while also supporting the proxy-level routing required to make the dynamic routes work.

Thanks!

**Reply by colmanhumphrey** [reply]
(don't think I can mark a comment as an answer - so marking the comment thread!)

**Reply by vgofaizen** [reply]
What we found is nextjs router pathname is not populated with the expected route on first hit if you use next export. So it does break it, @leerob . Until this issue is fixed within nextjs, one can use a provider in _app.js that wraps your components and adjusts the route or put this before the return statement inside you _app.js default function:
```
import { useRouter } from 'next/router'

const { asPath, push, pathname } = useRouter()
if (asPath.split('?')[0] != pathname.split('?')[0] && !pathname.includes('[')) {
    // Work around for next export breaking SPA routing on first hit
    console.log('Browser route ' + asPath + ' did not match nextjs router route ' + pathname)
    push(asPath)
    return <div>Loading...</div>
}
```
